src/storage-types/src/shim.rs

```diff
@@ -0,0 +1,543 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use super::errors::{proto_dataflow_error, ProtoUpsertValueError, ProtoUpsertValueErrorLegacy};
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum ProtoUpsertValueErrorShim {
+    New(ProtoUpsertValueError),
+    Legacy(Box<ProtoUpsertValueErrorLegacy>),
+}
+
+impl Default for ProtoUpsertValueErrorShim {
+    fn default() -> Self {
+        Self::New(ProtoUpsertValueError::default())
+    }
+}
+
+impl prost::Message for ProtoUpsertValueErrorShim {
+    fn encode_raw<B>(&self, buf: &mut B)
+    where
+        B: bytes::BufMut,
+        Self: Sized,
+    {
+        match self {
+            Self::New(new) => new.encode_raw(buf),
+            Self::Legacy(legacy) => legacy.encode_raw(buf),
+        }
+    }
+
+    fn merge_field<B>(
+        &mut self,
+        tag: u32,
+        wire_type: prost::encoding::WireType,
+        buf: &mut B,
+        ctx: prost::encoding::DecodeContext,
+    ) -> Result<(), prost::DecodeError>
+    where
+        B: bytes::Buf,
+        Self: Sized,
+    {
+        if tag == 1 {
+            // This is  the problematic tag that changed types. Here we need to do some type
+            // "inference" to figure out if we're decoding the new or the legacy format and update
+            // the variant of self accordingly.
+
+            // We need `merge_field` two times but we can only read the provided buffer once, so we
+            // make a local copy. This would normally be tricky with only the `Buf` bound but we
+            // know that we only get here as a result of a `Codec::decode` call, which has a &[u8]
+            // argument that is what B will be. This means that the `B: Buf` generic type is backed
+            // by a contiguous buffer and therefore calling `Buf::chunk` will give us all the data.
```

Comment:
> we only get here as a result of a `Codec::decode` call

That's not entirely true, right? `SourceDataDecode` calls `ProtoDataflowError::decode` directly. It passes a `&[u8]` so the rest of the reasoning still holds, but I'm a bit worried that we are overlooking over calls to `Message::decode` on one of the types that contain a shim. Is there any way we can assert that we really get all the data here?

Reply:
Alternatively, we could follow the simpler strategy of trying the decoding and `is_some` check for the new variant first, and if that fails decode with the old variant. AFAICT the only difference to the current behavior would be that things that successfully decode under both variants would not produce a panic but would be encoded with the new variant instead. Which seems fine?

Reply:
>  Is there any way we can assert that we really get all the data here?

Good idea, I added an assert that checks `data.len()` against `buf.remaining()`.

> Alternatively, we could follow the simpler strategy of trying the decoding and is_some check for the new variant first, and if that fails decode with the old variant

This isn't possible to do in here because the `buf` argument only implements the `Buf` trait, and you cannot use that trait without consuming its data. Were you thinking of something else?

Reply:
> Were you thinking of something else?

Nope, I just wasn't aware of that limitation.